src/charm.py

```diff
@@ -219,25 +279,130 @@ def _generate_kiali_layer() -> Layer:
             }
         )
 
-    def _get_prometheus_source_url(self):
+    def _get_grafana_uid(self) -> str:
+        """Return the UID of the related Grafana.
+
+        Raises:
+          ConfigurationBlockingError: If no grafana is related to this application
+          ConfigurationWaitingError: If a grafana is related to this application, but its data is incomplete.
+        """
+        if len(self._grafana_metadata) == 0:
+            raise ConfigurationBlockingError("No grafana available over the grafana-metadata relation")
```

Comment:
Instinctively, I dislike the idea that a method that's supposed to give you a grafana UID should have an opinion on whether, on failure, the charm should be put in this or that state or surface this or that status.
I think those should be separate concerns.

I'd rather have this method return Optional, and the caller decide what to do with it. In an event handler, this may mean proceeding or not with a configuration setup, in a collect-status handler, it may mean setting the one status or the other.

Reply:
I'd agree more with the argument that helpers shouldn't have an opinion on the caller's status if this was a separate object, but this is as helper of the charm class.  It is how *the Kiali Charm* fetches grafana's configuration - including an opinion about the problem seems in scope

Reply:
Renaming the exceptions, and having the collect-status handler map them to the status, feels more robust, yes.
As to the "we need more than Optional" to make that distinction: how so? 
This is imho the golden standard of the collect-status handlers: https://github.com/canonical/cos-lib/blob/main/src/cosl/coordinated_workers/coordinator.py#L599